evap/staff/importers.py

```diff
@@ -417,6 +470,93 @@ def process(cls, excel_content, test_run):
+
+        # since the user profiles are not necessarily saved to the database, they are not guaranteed to have a pk yet which
+        # makes anything relying on hashes unusable here (for a faster list difference)
+        self.users_to_add = list(filter(lambda x: x not in already_related, self.user_list))
+
+    def create_relations(self):
+        if self.import_type == 'participant':
+            self.course.participants.add(*self.users_to_add)
+            msg = _("{} Participants added to course {}").format(len(self.users_to_add), self.course.name)
+            msg += create_user_list_string_for_message(self.users_to_add)
+            self.success_messages.append(mark_safe(msg))
+
+        else:
+            for user in self.users_to_add:
+                order = Contribution.objects.filter(course=self.course).count()
+                contribution = Contribution(course=self.course, contributor=user, order=order)
+                contribution.save()
+                self.course.contributions.add(contribution)
+            msg = _("{} Contributors added to course {}").format(len(self.users_to_add), self.course.name)
+            msg += create_user_list_string_for_message(self.users_to_add)
+            self.success_messages.append(mark_safe(msg))
+
+    def create_test_success_messages(self):
+        if self.import_type == "participant":
+            msg = _("{} participants would be added to the course {}:").format(len(self.users_to_add), self.course.name)
+        else:  # import_type == "contributor"
+            msg = _("{} contributors would be added to the course {}:").format(len(self.users_to_add), self.course.name)
+        msg += create_user_list_string_for_message(self.users_to_add)
+
+        self.success_messages.append(mark_safe(msg))
+
+    def do_user_import(self, file_content, test_run):
+        self.user_list, self.success_messages, self.warnings, self.errors = UserImporter.process(file_content, test_run)
+
+    def get_users_from_source_course(self, source_course):
+        if self.import_type == 'participant':
+            self.user_list = source_course.participants.all()
+        else:
+            self.user_list = UserProfile.objects.filter(contributions__course=source_course)
+
+    @classmethod
+    def process(cls, import_type, course, test_run, file_content=None, source_course=None):
+        importer = cls(import_type, course)
+
+        if file_content:
+            importer.do_user_import(file_content, test_run)
+        elif source_course:
+            importer.get_users_from_source_course(source_course)
+        else:
+            raise ValueError("Expected one of the two positional arguments: file_content or source_course")
+
+        importer.handle_already_related()
+
+        if test_run:
+            importer.create_test_success_messages()
+        else:
+            importer.create_relations()
+
+        return importer.success_messages, importer.warnings, importer.errors
```

Comment:
every method in this class except this one branches on `if self.import_type == "participant":`, and the individual branches don't contain that much code. suggestion: do that branch exactly once and call one of two methods, e.g. `process_contributors` and `process_participants`, which have the code from the other methods in this class inlined, and using local variables instead of object attributes.